website/addons/box/model.py

```diff
@@ -0,0 +1,328 @@
+
+
+class BoxFile(GuidFile):
+    """A Box file model with a GUID. Created lazily upon viewing a
+    file's detail page.
+    """
+
+    #: Full path to the file, e.g. 'My Pictures/foo.png'
+    path = fields.StringField(required=True, index=True)
+
+    @property
+    def file_name(self):
+        if self.revision:
+            return '{0}_{1}_{2}.html'.format(self._id, self.revision, base64.b64encode(self.folder))
+        return '{0}_{1}_{2}.html'.format(self._id, self.unique_identifier, base64.b64encode(self.folder))
+
+    @property
+    def waterbutler_path(self):
+        path = self.path
+        if self.folder == '/':
+            return path
+        return path.replace(self.folder, '', 1)
+
+    @property
+    def folder(self):
+        return self.node.get_addon('box').folder
+
+    @property
+    def provider(self):
+        return 'box'
+
+    @property
+    def version_identifier(self):
+        return 'revision'
+
+    @property
+    def unique_identifier(self):
+        return self.path.split('/')[1][-5] #self._metadata_cache['extra']['revisionId']
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Get or create a new file record. Return a tuple of the form (obj, created)
+        """
+        cleaned_path = clean_path(path)
+        try:
+            new = cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', cleaned_path)
+            )
+            created = False
+        except ModularOdmException:
+            # Create new
+            new = cls(node=node, path=cleaned_path)
+            new.save()
+            created = True
+        return new, created
+
+
+class BoxUserSettings(AddonUserSettingsBase):
+    """Stores user-specific box information, including the Oauth access
+    token.
+    """
+
+    box_id = fields.StringField(required=False)
+    access_token = fields.StringField(required=False)
+    refresh_token = fields.StringField(required=False)
+    box_info = fields.DictionaryField(required=False)
+    token_type = fields.StringField(required=False)
+    restricted_to = fields.DictionaryField(required=False)
+    last_refreshed = fields.DateTimeField(editable=True)
+
+
+    # TODO(sloria): The `user` param in unnecessary for AddonUserSettings
+    def to_json(self, user=None):
+        """Return a dictionary representation of the user settings.
+        The dictionary keys and values will be available as variables in
+        box_user_settings.mako.
+        """
+        output = super(BoxUserSettings, self).to_json(self.owner)
+        output['has_auth'] = self.has_auth
+        return output
+
+    def token_refreshed_callback(self, access_token, refresh_token):
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.last_refreshed = datetime.utcnow()
+        self.save()
+
+    @property
+    def has_auth(self):
+        return bool(self.access_token and refresh_creds_if_necessary(self))
```

Comment:
Can `refresh_creds_if_necessary` be moved into this class as self is being passed to it anyways?